### PR TITLE
[AOT] Add CreateFunctionMetadata analysis pass

### PR DESCRIPTION
**Pull request by mbaret**
Tracking issue: https://github.com/apache/tvm/issues/12548

AOT requires FunctionInfo to be defined for all the functions in the module. This stores information on how much memory the functions use. This commit adds a separate analysis pass to create all the FunctionInfos + some tests for the new pass.

### CODE REVIEW

**Comment by mbaret** [comment]
cc also @ekalda 
(apologies for late follow-up, have been away for a while)

File: src/relay/backend/aot/create_function_metadata.cc
Comment (areusch) [inline src/relay/backend/aot/create_function_metadata.cc:58]: should we CHECK() here rather than letting `.value()` do it for us?

File: src/relay/backend/aot/create_function_metadata.cc
Comment (areusch) [inline src/relay/backend/aot/create_function_metadata.cc:61]: can you expand the var name? maybe `total_io_bytes` or something to map to the FunctionInfo member?

**Review by areusch — COMMENTED** [review]
thanks @mbaret , left a couple small comments.

**Review by lhutton1 — APPROVED** [review]
Thanks @mbaret, looks great! LGTM modulo @areusch's comments

**Review by lhutton1 — APPROVED** [review]
Nothing more from me, LGTM!

**Comment by mbaret** [comment]
@lhutton1 Andrew's on holiday for a couple of weeks, so would it be OK to merge without his approval? Think I addressed his comments :)

**Comment by lhutton1** [comment]
Thanks @mbaret @areusch!

**Pull request merged**
